File: LCAinBST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        qu = [root]

        while qu:
            curr = qu.pop()

            if curr == p or curr == q:
                return curr
            
            if p.val < curr.val < q.val or p.val > curr.val > q.val:
                return curr
            
            if curr.left:
                qu.append(curr.left)
            
            if curr.right:
                qu.append(curr.right)
        


            

File: LCA_in_BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        //make use of the bst proprty.

        return dfs(root, p, q);

    }

    public TreeNode dfs(TreeNode root, TreeNode p, TreeNode q){


        if(root == null){
            return null;
        }

        if(((p.val <= root.val) && q.val >= root.val) || ((q.val <= root.val) && p.val >= root.val)){
            return root;
        }

        //if both the nodes are greater than the current root then go to the right side of root. 

        if((p.val > root.val) && (q.val > root.val)){
            return dfs(root.right, p, q);
        }

        return dfs(root.left, p, q);
    }
}